src/components/SearchBar.js

```diff
@@ -0,0 +1,62 @@
+import { getSearchData } from "utils/data";
+import { dispatch } from "utils/dispatch";
+
+export class SearchBar {
+  searchBar = document.getElementById("search-bar");
+  autoCompleteElement = document.getElementById("search-bar-autocomplete");
+
+  constructor() {
+    this.searchBar.addEventListener("input", () => {
+      dispatch.call("search-bar-autocomplete", this, this.searchBar.value);
+    });
+    this.searchBar.addEventListener("blur", () => {
+      dispatch.call("remove-autocompete-dropdown", this);
+    });
+    this.searchBar.addEventListener("focus", () => {
+      this.searchBar.value = "";
+    });
+    dispatch.on("search-bar-autocomplete", this.autoComplete);
+    dispatch.on("remove-autocompete-dropdown", this.removeAutocomplete);
+    dispatch.on("search-fetch-data-reject", (err) => console.error(err));
+  }
+
+  removeAutocomplete() {
+    setTimeout(() => {
+      this.autoCompleteElement.innerHTML = "";
+      this.searchBar.value =
+        "\u{01f50d}  Enter Nation, state, province, city or zipcode";
+    }, 400);
+  }
+
+  async autoComplete(str) {
+    if (str.length > 3) {
+      const res = await getSearchData(str);
+      const features = res.features.filter(
+        (feature) => feature.bbox !== undefined
+      );
+      this.autoCompleteElement.innerHTML = features
+        .map((feature) => this.autocompleteElement(feature))
+        .join("");
+      features.forEach((feature) => {
+        document
+          .getElementById(feature.id)
+          .addEventListener("click", () =>
+            dispatch.call(
+              "choose-autocomplete-element",
+              Window.lmap,
```

Comment:
Just FYI it's okay to leave the `this` argument for `dispatch.call()` as undefined. You can get around fixing the context of `this` in `LeafletMap` by making the callback an arrow function.

Reply:
This did not work for me. When I left the `this` argument undefined, then `this` became the caller instead of the callee.